src/ipoplib.py

```diff
@@ -626,21 +747,46 @@ def parse_config():
     if not ("xmpp_username" in CONFIG and "xmpp_host" in CONFIG):
         raise ValueError("At least 'xmpp_username' and 'xmpp_host' must be "
                          "specified in config file or string")
-
+    # Use custom keyring incase of non-interactive environment
+    # (uses a fixed password)
+    if isinstance(keyring.get_keyring(),keyring.backends.file.EncryptedKeyring):
+        keyring.set_keyring(IPOPEncryptedKeyring())
+    save_password = False
     if not args.update_config:
-        temp = keyring.get_password("ipop", CONFIG["xmpp_username"])
-    if temp == None and "xmpp_password" not in CONFIG:
-        prompt = "\nPassword for %s: " % CONFIG["xmpp_username"]
-        if args.pwdstdout:
-          CONFIG["xmpp_password"] = getpass.getpass(prompt, stream=sys.stdout)
+        if "xmpp_password" in CONFIG:
+            # password is present in config file. No need to look
+            # for other options store password in keyring
+            temp = CONFIG["xmpp_password"]
+            save_password = True
+            # we need to store it in keyring because there is no way
+            # to distinguish between environments where password is 
+            # always present in config file and otherwise. So even though 
+            # config with password is going to be used always, the keyring 
+            # might prompt for password for keyring on first run
         else:
-          CONFIG["xmpp_password"] = getpass.getpass(prompt)
-    if temp != None:
-        CONFIG["xmpp_password"] = temp
-    try:
-        keyring.set_password("ipop", CONFIG["xmpp_username"],CONFIG["xmpp_password"])
-    except:
-        raise RuntimeError("Unable to store password in keyring")
+            # Try to retrieve password from keyring
+            temp = keyring.get_password("ipop", CONFIG["xmpp_username"])
+            # Try to request valid password from user
+            while temp == None or temp == "":
+                 prompt = "\nPassword for %s: " % CONFIG["xmpp_username"]
+                 if args.pwdstdout:
+                      temp = getpass.getpass(prompt, stream=sys.stdout)
+                 else:
+                      temp = getpass.getpass(prompt)
+                 save_password = True
+        if temp != None:
+            CONFIG["xmpp_password"] = temp
+    else:
+        save_password = True
+     
+    if save_password:
+        # Save password in keyring
+        try:
+            keyring.set_password("ipop", CONFIG["xmpp_username"],\
+                                 CONFIG["xmpp_password"])
+        except Exception, err:
+            print(traceback.format_exc())
```

Comment:
In particular, `logging.error(msg, exc_info=True)`, so you have no need to catch `Exception`, nor name it.